### PR TITLE
Timeout `GreedyPauliSimp`

### PR DESCRIPTION
**Pull request by sjdilkes**
Add `trials` option to `GreedyPauliSimp`, running multiple attempts at `GreedyPauliSimp` and returning the "smallest" circuit.

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: tket/src/Transformations/GreedyPauliOptimisation.cpp [hunk 1]
```diff
@@ -775,26 +811,71 @@ Circuit greedy_pauli_graph_synthesis(
...
-        std::future_status::ready) {
-      circ = future.get();
-      circ.decompose_boxes_recursively();
-      return true;
+                    max_tqe_candidates, seed, allow_zzphase, thread_timeout,
+                    trials](Circuit& circ) {
+    std::mt19937 seed_gen(seed);
+    std::queue<
+        std::pair<std::future<Circuit>, std::shared_ptr<std::atomic<bool>>>>
+        all_threads;
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/src/Transformations/GreedyPauliOptimisation.cpp:837]: not anymore!

File: pytket/binders/transform.cpp [hunk 2]
```diff
@@ -440,12 +440,14 @@ PYBIND11_MODULE(transform, m) {
           "\n:param allow_zzphase: If set to True, allows the algorithm to "
           "implement 2-qubit rotations using ZZPhase gates when deemed "
           "optimal. Defaults to False."
-          "\n:param timeout: Sets maximum out of time spent finding solution."
+          "\n:param thread_timeout: Sets maximum out of time spent finding a "
+          "single solution in one thread."
+          "\n:param trials: Sets maximum number of found solutions."
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply pytket/binders/transform.cpp:456]: Yeah good point, have done

File: pytket/binders/transform.cpp [hunk 3]
```diff
@@ -451,12 +451,18 @@ PYBIND11_MODULE(transform, m) {
...
           "implement 2-qubit rotations using ZZPhase gates when deemed "
           "optimal. Defaults to False."
-          "\n:param timeout: Sets maximum out of time spent finding solution."
+          "\n:param thread_timeout: Sets maximum out of time spent finding a "
+          "single solution in one thread."
+          "\n:param trials: Sets maximum number of found solutions. The "
+          "circuit "
+          "smallest circuit is returned, priorising the number of 2qb-gates, "
+          "then "
+          "gates, then depth."
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply pytket/binders/transform.cpp]: done!

**Comment by yao-cqc** [comment]
Still need to update the stubs

**Comment by sjdilkes** [comment]
> Still need to update the stubs

Oops done, sorry forgot to push

**Review by yao-cqc — APPROVED** [review]

**Pull request merged**